samples/drivers/flash_shell/src/main.c

```diff
@@ -227,6 +248,140 @@ static int do_write(const struct shell *shell, off_t offset, uint8_t *buf,
 	return ret;
 }
 
+static int do_write_unaligned(const struct shell *shell, off_t offset, uint8_t *buf,
+		    size_t len, bool read_back)
+{
+	int ret;
+	size_t page_size = flash_get_write_block_size(flash_device);
+	size_t size_before = offset % page_size;
+	size_t size_after = page_size - ((size_before + len) % page_size);
+	size_t aligned_size = size_before + len + size_after;
+	off_t  start_page = offset - size_before;
+	off_t  last_page = start_page + aligned_size - page_size;
+	bool single_page_write = (size_before + len < page_size);
+
+	static char *before_data;
+	static char *after_data;
+
+	ret = flash_write_protection_set(flash_device, false);
+	if (ret) {
+		PR_ERROR(shell, "Failed to disable flash protection (err: %d)."
+				"\n", ret);
+		return ret;
+	}
+
+	if (0 == size_before && 0 == size_after) {
+		/* Aligned write */
+		flash_erase(flash_device, offset, len);
+		flash_write(flash_device, offset, buf, len);
+		ret = flash_write_protection_set(flash_device, true);
+		if (ret) {
+			PR_ERROR(shell, "Failed to enable flash protection (err: %d)."
+					"\n", ret);
+		}
+		return ret;
+	}
+
+	if (!before_data) {
+		/* Allocate once */
+		before_data = k_malloc(page_size);
+		after_data = k_malloc(page_size);
+
+		if (!before_data || !after_data) {
```

Comment:
If `before_data` allocation works, but `after_data` fails, this will leak the allocation of `before_data`.